File: recipes/subms-adaptive-radix-tree/java/src/main/java/com/submillisecond/recipes/art/features/RangeScan.java
package com.submillisecond.recipes.art.features;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import com.submillisecond.recipes.art.Art;
import com.submillisecond.recipes.art.ArtInternals;

/**
 * In-order scan between two byte-lex bounds.
 *
 * <p>Ordering is byte-lexicographic. Two consequences worth naming up
 * front:
 * <ul>
 *   <li>UTF-8 strings compare the same way as their UTF-8 byte
 *       sequences in this scheme. For ASCII-only keys the order matches
 *       what a reader would expect from a dictionary. Non-ASCII
 *       multi-byte sequences sort by their codepoint encoding, which is
 *       NOT a locale-aware collation.</li>
 *   <li>Empty key is the minimum element.</li>
 * </ul>
 *
 * <p>Bounds are independently inclusive / exclusive / unbounded.
 *
 * <p>Byte-equivalent to the Rust sibling
 * {@code subms_adaptive_radix_tree::features::range_scan}.
 */
public final class RangeScan {

    private RangeScan() {}

    public enum BoundKind { UNBOUNDED, INCLUDED, EXCLUDED }

    public static final class Bound {
        public final BoundKind kind;
        public final byte[] value;
        private Bound(BoundKind kind, byte[] value) {
            this.kind = kind;
            this.value = value;
        }
        public static Bound unbounded() { return new Bound(BoundKind.UNBOUNDED, null); }
        public static Bound included(byte[] value) { return new Bound(BoundKind.INCLUDED, value); }
        public static Bound excluded(byte[] value) { return new Bound(BoundKind.EXCLUDED, value); }
    }

    public static final class Entry<V> {
        public final byte[] key;
        public final V value;
        public Entry(byte[] key, V value) {
            this.key = key;
            this.value = value;
        }
    }

    public static <V> List<Entry<V>> range(Art<V> tree, Bound from, Bound to) {
        List<byte[]> keys = new ArrayList<>();
        List<V> values = new ArrayList<>();
        ArtInternals.collect(tree, keys, values);
        // collect() already produces byte-lex order; filter to bounds.
        List<Entry<V>> out = new ArrayList<>();
        Comparator<byte[]> cmp = RangeScan::lex;
        for (int i = 0; i < keys.size(); i++) {
            byte[] k = keys.get(i);
            boolean fromOk = switch (from.kind) {
                case UNBOUNDED -> true;
                case INCLUDED -> cmp.compare(k, from.value) >= 0;
                case EXCLUDED -> cmp.compare(k, from.value) > 0;
            };
            if (!fromOk) continue;
            boolean toOk = switch (to.kind) {
                case UNBOUNDED -> true;
                case INCLUDED -> cmp.compare(k, to.value) <= 0;
                case EXCLUDED -> cmp.compare(k, to.value) < 0;
            };
            if (!toOk) continue;
            out.add(new Entry<>(k, values.get(i)));
        }
        return out;
    }

    private static int lex(byte[] a, byte[] b) {
        int len = Math.min(a.length, b.length);
        for (int i = 0; i < len; i++) {
            int ai = a[i] & 0xff;
            int bi = b[i] & 0xff;
            if (ai != bi) return Integer.compare(ai, bi);
        }
        return Integer.compare(a.length, b.length);
    }
}

File: recipes/subms-adaptive-radix-tree/rust/src/features/range_scan.rs
//! In-order scan between two byte-lex bounds.
//!
//! Ordering is byte-lexicographic. Two consequences worth naming up
//! front:
//!
//! - UTF-8 strings compare the same way as their UTF-8 byte sequences
//!   in this scheme. For ASCII-only keys the order matches what a
//!   reader would expect from a dictionary. Non-ASCII multi-byte
//!   sequences sort by their codepoint encoding, which is NOT a
//!   locale-aware collation - if you need collation, sort the result
//!   externally.
//! - Empty key is the minimum element.
//!
//! Bounds are independently inclusive / exclusive / unbounded.

use crate::{Art, Node};

#[derive(Clone)]
pub enum Bound<'a> {
    Included(&'a [u8]),
    Excluded(&'a [u8]),
    Unbounded,
}

/// In-order scan. Returns `(key, &value)` pairs sorted by key (byte-lex).
///
/// Allocates one `Vec` for the result and one for each visited path,
/// which is acceptable for cookbook clarity. A streaming iterator would
/// avoid the result Vec at the cost of significantly hairier lifetimes.
pub fn range<'a, V>(tree: &'a Art<V>, from: Bound<'a>, to: Bound<'a>) -> Vec<(Vec<u8>, &'a V)> {
    let mut out: Vec<(Vec<u8>, &'a V)> = Vec::new();
    let mut prefix: Vec<u8> = Vec::new();
    walk(tree.root(), &mut prefix, &from, &to, &mut out);
    out
}

fn walk<'a, V>(
    node: &'a Node<V>,
    prefix: &mut Vec<u8>,
    from: &Bound<'a>,
    to: &Bound<'a>,
    out: &mut Vec<(Vec<u8>, &'a V)>,
) {
    // Path compression: the node's prefix bytes precede its value + edges.
    prefix.extend_from_slice(&node.prefix);
    if let Some(v) = node.value.as_ref() {
        if in_bounds(prefix, from, to) {
            out.push((prefix.clone(), v));
        }
    }

    for (byte, child) in node.children.sorted_pairs() {
        prefix.push(byte);
        // Early-skip subtrees that can't contain a key in range. `prefix` is a
        // lower bound on every key below, so the prune stays sound with
        // compression (the child adds only more bytes).
        if !subtree_can_overlap(prefix, from, to) {
            prefix.pop();
            continue;
        }
        walk(child, prefix, from, to, out);
        prefix.pop();
    }
    prefix.truncate(prefix.len() - node.prefix.len());
}

fn in_bounds(key: &[u8], from: &Bound, to: &Bound) -> bool {
    let from_ok = match from {
        Bound::Unbounded => true,
        Bound::Included(b) => key >= *b,
        Bound::Excluded(b) => key > *b,
    };
    let to_ok = match to {
        Bound::Unbounded => true,
        Bound::Included(b) => key <= *b,
        Bound::Excluded(b) => key < *b,
    };
    from_ok && to_ok
}

/// Could the subtree rooted at `prefix` contain a key in `[from, to]`?
/// We can prune by the *upper* bound (no key in the subtree can be less
/// than `prefix`, but they can grow arbitrarily larger). For the lower
/// bound we can NOT prune via the prefix alone because a longer
/// descendant key may still satisfy `key >= from`. So we only check the
/// upper bound here.
fn subtree_can_overlap(prefix: &[u8], _from: &Bound, to: &Bound) -> bool {
    match to {
        Bound::Unbounded => true,
        Bound::Included(b) => prefix <= as_slice(b, prefix.len()),
        Bound::Excluded(b) => {
            let trimmed = as_slice(b, prefix.len());
            // For Excluded(b) the subtree is still in range as long as
            // prefix isn't already past b's truncation; we allow ==.
            prefix <= trimmed
        }
    }
}

fn as_slice(b: &[u8], len: usize) -> &[u8] {
    if b.len() >= len { &b[..len] } else { b }
}

#[cfg(test)]
#[path = "range_scan_tests.rs"]
mod tests;
